acvm/src/lib.rs

```diff
@@ -222,13 +250,18 @@ pub trait ProofSystemCompiler {
 
     /// Generates a proving and verification key given the circuit description
     /// These keys can then be used to construct a proof and for its verification
-    fn preprocess(&self, circuit: &Circuit) -> Result<(Vec<u8>, Vec<u8>), Self::Error>;
+    fn preprocess(
```

Comment:
I think preprocess() can be confusing with preprocessing snarks. Normally it is the generator which computes the keys, so I would instead have these 2 functions:
generate_proving_key(crs, circuit)
generate_verification_key(crs, circuit)

Is there an advantage to do both at the same time?

Reply:
I'm fine with that, but I'd split it into a different PR. Pinging @vezenovm because he added preprocess, I believe.

Reply:
> Is there an advantage to do both at the same time?

To my knowledge proving schemes generally generate the verification key based upon the proving key (at least this is how it is done in barretenberg). Looking into the method @phated just linked, the method actually fetches the vk from the pk [internally](https://github.com/arkworks-rs/groth16/blob/9bac46fb3b1fbc0d2c88c6bc59b0ba798216bf6f/src/lib.rs#L71). Separating the preprocess method could lead to situations that require generating the proving key twice (which is the most intensive part of the preprocess operation) without making API changes to the proving system. Of course we don't want to base our API off of proving systems and rather those looking to integrate should change their API instead. However, I don't see the disadvantage of having preprocess be one method.

In our `aztec_backend` we have two separate composer methods to compute the proving key and the verification key (which takes in a proving key) so our API technically could support `generate_proving_key` and `generate_verification_key` without a performance degradation, but this does not look to be the case in all proving schemes. Having one `preprocess` method covers the case where a proving system supports generating the proving key and verification key separately, as well as at the same time. 

I would support adding two functions `generate_proving_key` and `generate_verification_key` as part of ACVM API, while keeping `preprocess`. The backend system could then chose if it wants to integrate these methods separately or just a single `preprocess` method. Either way I think this is entirely separate from the reference string work in this PR and should be a separate PR as it requires a new API change.

Reply:
I separated them because a verifier just need the verification key.
However, if in practise they are always computed at the same time, then I guess it is ok to have a generate() method, and may be some get_proving_key() and get_verification_key().

Reply:
Yeah I think we can discussing adding two `generate_proving_key` and `generate_verification_key` methods, but it is still a different scope than this PR

Reply:
Since this API design is out of the scope of this PR, I opened an issue to continue the discussion: #256